### ISSUE TITLE
Consider adding finalize/dispose actions

### ISSUE THREAD

**Issue by paulomorgado** [simpleinjector/SimpleInjector#519]
We can register an initializer action for registered types.

It would be nice to register an action for when the container is disposed.

Sometimes I have disposable objects that are only on instance creators that I would like to have disposed when the container disposes its disposables.

**Comment by dotnetjunkie** [comment]
There are multiple ways to achieve this. Simplest way is to create a list of the disposable instances and dispose them once you dispose the container. To make this easier, you can create a wrapper class:

``` c#
public sealed class DisposableComposite : List<IDisposable>, IDisposable
{
    private readonly List<IDisposable> disposables = new List<IDisposable>();

    public void Add(IDisposable disposable) {
        lock (this.disposables) {
            this.disposables.Add(disposable);
        }
    }

    public void Dispose() => this.disposables.ForEach(d => d.Dispose());
}
```

Usage:

``` c#
var containerDisposables = new DisposableComposite();
var container = new Container();

container.RegisterInitializer(instance =>
{
    containerDisposables.Add(instance.Dep = new Disposable());
});

container.Dispose();
RegisterInitializer.Dispose();
```

You can also integrate this into the Container by deriving from the container:

``` c#
public sealed class DisposableContainer : Container
{
    private readonly containerDisposables = new DisposableComposite();
    public void RegisterForDisposal(IDisposable d) => containerDisposables.Add(d);
    override protected void Dispose(bool disposing) {
        try {
            base.Dispose(disposing);
        } finally {
            if (disposing) container.Disposables.Dispose();
        }
    }
}
```

Usage:

``` c#
var container = new DisposableContainer();

container.RegisterInitializer(instance =>
{
    container.AddForDisposal(instance.Dep = new Disposable());
});

container.Dispose();
```

You can also register the `DisposableComposite` as singleton in the container. The container will automatically dispose it, when it itself is disposed. Admittedly, this is a bit tricky, because you will have to register it using the lambda overload of `RegisterSingleton<T>(Func<T>)`, and you will have to make sure this instance is resolved (at least) once from the container, to make sure the container will internally add it to its `disposableSingletonsScope Scope` instance. You need to use `RegisterSingleton<T>(Func<T>)` because `RegisterSingleton<T>(T)` will _not_ dispose supplied instances.

**Comment by paulomorgado** [comment]
Will just registering `T` with `Func\<T\>´ and not retrieving `T` at any point call `Dispose` on `T` when the container is disposed?

**Comment by dotnetjunkie** [comment]
> Will just registering T with Func<T> and not retrieving T at any point call Dispose on T when the container is disposed?

If you do this:

``` c#
var container = new Container();
var d = new MyDisposable();
container.RegisterSingleton(() => d);
container.Dispose();
```

In this case, `d` will **not** get disposed. 

In the following case however, it _will_ get disposed:


``` c#
var container = new Container();
var d = new MyDisposable();
container.RegisterSingleton(() => d);
container.Verify(); // verify ensures that the Func<MyDisposable> is called.
container.Dispose();
```

**Comment by dotnetjunkie** [comment]
I'll keep this issue open. This is something we might want to simplify in the future. An issue solution for instance is to allow access to the container's internal `Scope` that is used to hold the disposable singletons. When the user has access to that, it will be trivial to register instances for disposal. For instance:

``` c#
container.SingletonScope.RegisterForDisposal(myDisposable);
```

**Comment by dotnetjunkie** [comment]
Moved this feature here #664.

**Comment by dotnetjunkie** [comment]
@paulomorgado, I just released [4.5.0-beta1](https://github.com/simpleinjector/SimpleInjector/releases/tag/v4.5.0-beta1) that gives access to the singleton `Scope` through the `Container.ContainerScope` property. Feel free to check this beta release out and give feedback on that feature (now you still can).
